assignment6/src/Parking.java

```diff
@@ -0,0 +1,107 @@
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.logging.Logger;
+
+import static java.util.Collections.nCopies;
+
+/**
+ * Parking class for rearranging cars
+ */
+public class Parking {
+    /**
+     * Return sequence of movement cars (from which places cars move to which).
+     * Function changes beginParking.
+     * <p>
+     * Complexity: O(n)
+     *
+     * @param beginParking -- sequence of cars on the beginning (from zero)
+     * @param endParking   -- sequence of cars on the end
+     * @return -- sequence of movements
+     */
+    public List<Movement> RearrangingCars(int[] beginParking, int[] endParking) {
+        List<Integer> beginCarsPosition = createCarsPositionList(beginParking);
+        List<Movement> movements = new LinkedList<>();
+
+        while (true) {
+            int posOfZero = beginCarsPosition.get(0);
+            int carToPosOfZero = endParking[posOfZero];
+            int posOfCarToZeroPos = beginCarsPosition.get(carToPosOfZero);
+
+            if (endParking[posOfZero] != 0) {
+                makeMovement(posOfZero, posOfCarToZeroPos, beginCarsPosition, beginParking, movements);
+            } else {
+                int posCarToZero = firstPosDifferentCar(beginParking, endParking);
+                if (posCarToZero == -1) {
+                    break;
```

Comment:
Using "while(true)" and "break" is a little bit ugly, but given the way the loop's structured, I'll admit I can't see an easy and pleasant way to arrange it differently. Ditto use of a special distinguished value (-1) - ugly, but the easiest way to do it in this context.